ros2doctor/ros2doctor/api/__init__.py

```diff
@@ -12,22 +12,109 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from typing import List
+from typing import Set
+from typing import Tuple
+
 from pkg_resources import iter_entry_points
 
+from ros2doctor.api.format import doctor_warn
+
+
+class DoctorCheck:
+    """Abstract base class of ros2doctor check."""
+
+    def category(self) -> str:
+        """:return: string linking checks and reports."""
+        raise NotImplementedError
+
+    def check(self) -> bool:
+        """:return: boolean indicating result of checks."""
+        raise NotImplementedError
+
+
+class DoctorReport:
+    """Abstract base class of ros2doctor report."""
+
+    def category(self) -> str:
+        """:return: string linking checks and reports."""
+        raise NotImplementedError
+
+    def report(self) -> 'Report':  # using str as wrapper for custom class Report
+        """:return: Report object storing report content."""
+        raise NotImplementedError
 
-def run_checks():
-    """Run all checks when `ros2 doctor/wtf` is called."""
-    all_results = []
-    failed_names = []
-    for check in iter_entry_points('ros2doctor.checks'):
-        result = check.load()()  # load() returns method
-        all_results.append(result)
+
+class Report:
+    """Stores report name and content."""
+
+    __slots__ = ['name', 'items']
+
+    def __init__(self, name: str):
+        """Initialize with report name."""
+        self.name = name
+        self.items = []
+
+    def add_to_report(self, item_name: str, item_info: str) -> None:
+        """Add report content to items list (list of string tuples)."""
+        self.items.append((item_name, item_info))
+
+
+def run_checks() -> Tuple[Set[str], int, int]:
+    """
+    Run all checks and return check results.
+
+    :return: 3-tuple (categories of failed checks, number of failed checks,
+             total number of checks)
+    """
+    failed_cats = set()  # remove repeating elements
+    fail = 0
+    total = 0
+    for check_entry_pt in iter_entry_points('ros2doctor.checks'):
+        try:
+            check_class = check_entry_pt.load()
+        except ValueError:
```

Comment:
It looks like `load()` can raise `UnknownExtra` (`from pkg_resources import UnknownExtra`) or `ImportError`. I don't see `ValueError` in the docs.

https://setuptools.readthedocs.io/en/latest/pkg_resources.html#entrypoint-objects